### ISSUE TITLE
A few questions for the usage

### ISSUE THREAD

**Issue by jinserk** [graykode/matorage#1]
It's really fantastic! Thank you so much for sharing this project.
I had a quick test with `minio` docker process and confirmed it works really well as expected.
I'd like to ask a few questions about the usage:
- can I save any objects other than tensors, i.e. a tuple of tensors, a dict or a sparse tensors? In this case how to give the attributes?
- If I add some more data samples to an existing dataset (in the case that periodically data samples could be added so if I have to refresh all the datasets with the added samples), will it be okay to add to the dataset and save it?
- If I use this in a distributed training, will the each dataloader occupy the amount of dataset? For example, if I'd like to use this with PyTorch DDP, then each process will have its own `DataLoader` and will load the data samples during the training -- so I wonder the loaded dataset size will be multiple of the original dataset size per process or not.

**Comment by graykode** [comment]
@jinserk Thank you for your interest in the project!

- can I save any objects other than tensors, i.e. a tuple of tensors, a dict or a sparse tensors? In this case how to give the attributes?: As can be seen in the existing example, tensors in the form of tuple or dict can also be stored. We do not define a new operation for sparse sensors, but we can save it. As for the tensor for sparse, we will add it to the long-term plan. Below is an example of storage in the form of a tuple :
```python
attributes=[
        ('image', 'float32', (1, 28, 28)),
        ('target', 'int64', (1))
    ]

traindata_saver({
        'image': image,
        'target': target
    })
```
- If I add some more data samples to an existing dataset (in the case that periodically data samples could be added so if I have to refresh all the datasets with the added samples), will it be okay to add to the dataset and save it? : If you simply add more data (append mode) does not matter if you save using the existing config. However, if you want to refresh your data, it is not currently implemented in code. If you want to refresh dataset (this means the same as you want to remove buckets), you should use minio web site or [mc command](https://github.com/minio/mc) from minio. (`mc rb --force --dangerous local/<bucket_name>`). I will further implement the refresh method by adding an aggregate on datasaver like this:
```python
traindata_saver({
        'image': image,
        'target': target
    }, refresh=True) # I will add this argument
```
- If I use this in a distributed training, will the each dataloader occupy the amount of dataset? For example, if I'd like to use this with PyTorch DDP, then each process will have its own DataLoader and will load the data samples during the training -- so I wonder the loaded dataset size will be multiple of the original dataset size per process or not. :I understood this question as follows: With distributed data parallel (DDP), N data is replicated as many nodes as M, and the total number of N*M data is not generated? Yes, that's right. When dataset of `torch.matorage` is initialized, it goes through a logic to pre-download data corresponding to config (https://github.com/graykode/matorage/blob/master/matorage/data/data.py#L80). However, this is very inefficient, So we recommend using the network access storage (NAS) option when using DDP. No new downloads of data are available when using NAS. Therefore, N datasets can be kept intact and DDP train can be performed.
- It looks like supporting only PyTorch and TensorFlow now, but how about the numpy array or matrix for scikit-learn or XGBoost? Can I store some numpy objects as well? : When the sensors in Pytorch and Tensorflow are saved, they are converted to numpy and saved. As with tfrecord, the tensor can be encoded in the proto-buf format, but because of its lack of universality, the most popular numpy format was used. Therefore, if the storage format is numpy array, you can save both scikit-learn or XGBOost. But of course, when it comes to data loaders, new additional implementations are needed.

Best Regard Tae Hwan

**Comment by jinserk** [comment]
@graykode Thank you very much for the detailed answers! It's really helpful, and very impressive.

My first question is actually related to heterogeneous shape of tensors, which means the case that the image size of your example could be changed sample by sample. Practically I'm working on a chemical problem -- molecule classification in chemistry or pharma companies -- and the input feature can be a graph which sizes are various according to the molecules. I know this cannot be simply implemented using an attributes, and that's why I asked about the sparse matrices support. Hope this could be implemented and be used sooner! :)

**Comment by graykode** [comment]
@jinserk 
Matrices with atypical shapes are difficult to store regardless of sparse. Moreover, sparse is not difficult to implement because it is guaranteed by existing scipy and hdf5. However, it is very difficult to store a tensor with an undefined shape as hdf5.

I recommend setting the maximum shape of the tensor of the heterogeneous shape and allowing it to be stored within that frame. If the range is very large, it is recommended to manage by dividing the frame (e.g. 1-5 is 5, 6-10 is 10)

**Comment by jinserk** [comment]
Good question. Basically I'm using a fixed shape for the input of the model. In the training, I just pad the heterogeneous shapes of the input with a fixed shape of max dim values.

**Comment by graykode** [comment]
If so, how about storing the fixed tensor itself that goes into the model's input in matorage? This is the core idea of matorage.

**Comment by jinserk** [comment]
Thanks for the suggestion, @graykode ! I will try to do so, since I don't know how the "compression" will work well. I had once tried to store the fixed tensors but the serialized file was almost 400 GB (used `torch.save`) while the file with sparse tensor was only 4 GB. I still hope the storing sparse tensor will be supported in this `matorage` project soon. :)

**Comment by graykode** [comment]
@jinserk 

I understand that there is currently no official support for hdf5 for sparse matrices. Therefore, the [official pytable document also recommends compressing for sparse matrix](http://www.pytables.org/usersguide/filenode.html?highlight=sparse).

In fact, according to many sources, it is recommended to use compression when using the hdf5 format. (https://stackoverflow.com/a/25678471/5350490) According to this, since it is a sparse matrix, **the original 512MB of data can be compressed up to 4.5KB.** So, can you experiment with your 400GB data and give you the final compressed result size? Try to `compression='gzip' with level=9`
